pkg/controller/machineset.go

```diff
@@ -677,6 +694,75 @@ func (c *controller) terminateMachines(inactiveMachines []*v1alpha1.Machine, mac
 	return nil
 }
 
+func (c *controller) terminateStaleMachines(inactiveMachines []*v1alpha1.Machine, deletionWindow int, machineSet *v1alpha1.MachineSet) error {
+	var (
+		wg    sync.WaitGroup
+		errCh = make(chan error, deletionWindow)
+	)
+	defer close(errCh)
+
+	wg.Add(deletionWindow)
+	for i := 0; i < deletionWindow; i++ {
+		go c.prepareMachineForDeletion(inactiveMachines[i], machineSet, &wg, errCh)
+	}
+	wg.Wait()
+
+	select {
+	case err := <-errCh:
+		// all errors have been reported before and they're likely to be the same, so we'll only return the first one we hit.
+		if err != nil {
+			return err
+		}
+	default:
+	}
+
+	return nil
+}
+
+// calculateStaleMachineDeletionWindow computes
+func (c *controller) calculateStaleMachineDeletionWindow(activeMachines, runningMachines, inactiveMachines []*v1alpha1.Machine, machineSet *v1alpha1.MachineSet) int {
+	var (
+		deletionWindow          int
+		numOfActiveMachines     = len(activeMachines)
+		numOfRunningMachines    = len(runningMachines)
+		numOfInactiveMachines   = len(inactiveMachines)
+		numOfMachineSetReplicas = int(machineSet.Spec.Replicas)
+	)
+
+	// nothing to do
+	if numOfInactiveMachines == 0 {
+		return 0
+	}
+
+	// Will not delete any stale machines. It is expected that the user deletes these machines manually.
+	if c.safetyOptions.FailedMachineDeletionRatio == 0 {
+		return 0
+	}
+
+	if c.safetyOptions.FailedMachineDeletionRatio == 1 {
+		return numOfInactiveMachines
+	}
+
+	if numOfMachineSetReplicas == 0 {
+		deletionWindow = numOfInactiveMachines
+	} else {
+		deletionWindow = int(math.Round(float64(numOfMachineSetReplicas) * c.safetyOptions.FailedMachineDeletionRatio))
+		if deletionWindow == 0 {
+			deletionWindow = 1
+		}
+
+		pendingMachines := numOfActiveMachines - numOfRunningMachines
```

Comment:
numOfActiveMachines (Pending, Available, Running, Unknown) - numOfRunningMachines (Running) = Machines which are either of the following statuses - `Available, Running, Unknown`? So it's not just pending machines, right?

Reply:
I've expanded on the issue in [another thread](https://github.com/gardener/machine-controller-manager/pull/482#discussion_r468195769). Let's discuss the naming and the algorithm there.